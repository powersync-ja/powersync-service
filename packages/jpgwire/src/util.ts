import * as tls from 'tls';
import { DEFAULT_CERTS } from './certs.js';
import * as pgwire from './pgwire.js';
import { PgType } from './pgwire_types.js';
import * as datefns from 'date-fns';

// TODO this is duplicated, but maybe that is ok
export interface NormalizedConnectionConfig {
  id: string;
  tag: string;

  hostname: string;
  port: number;
  database: string;

  username: string;
  password: string;

  sslmode: 'verify-full' | 'verify-ca' | 'disable';
  cacert: string | undefined;

  client_certificate: string | undefined;
  client_private_key: string | undefined;
}

type Mutable<T> = {
  -readonly [P in keyof T]: T[P];
};

export interface PgWireConnectionOptions extends NormalizedConnectionConfig {
  resolved_ip?: string;
}

export function clientTlsOptions(options: PgWireConnectionOptions): tls.ConnectionOptions {
  if (options.client_certificate && options.client_private_key) {
    return {
      cert: options.client_certificate,
      key: options.client_private_key
    };
  } else {
    return {};
  }
}
export function tlsOptions(options: PgWireConnectionOptions): false | tls.ConnectionOptions {
  if (options.sslmode == 'disable') {
    return false;
  } else if (options.sslmode == 'verify-full') {
    // Validate against default CA, or supplied CA.
    // Also validate hostname.
    return {
      ca: options.cacert ?? DEFAULT_CERTS,
      // hostname for TLS validation.
      // This may be different from the host we're connecting to if we pre-resolved
      // the IP.
      host: options.hostname,
      servername: options.hostname,
      ...clientTlsOptions(options)
    };
  } else if (options.sslmode == 'verify-ca') {
    if (options.cacert == null) {
      throw new Error(`cacert required for verify-ca`);
    }
    return {
      ca: options.cacert!,
      host: options.hostname,
      servername: options.hostname,
      // Disable hostname check
      checkServerIdentity: () => undefined,
      ...clientTlsOptions(options)
    };
  } else {
    throw new Error(`Unsupported sslmode: ${options.sslmode}`);
  }
}

export async function connectPgWire(config: PgWireConnectionOptions, options?: { type?: 'standard' | 'replication' }) {
  let connectionOptions: Mutable<ExtendedPgwireOptions> = {
    application_name: 'PowerSync',

    // tlsOptions below contains the original hostname
    hostname: config.resolved_ip ?? config.hostname,
    port: config.port,
    database: config.database,

    user: config.username,
    password: config.password
  };

  if (options?.type != 'standard') {
    connectionOptions.replication = 'database';
  }

  if (config.sslmode != 'disable') {
    connectionOptions.sslmode = 'require';

    // HACK: Not standard pgwire options
    // Just the easiest way to pass on our config to pgwire_node.js
    connectionOptions.sslrootcert = tlsOptions(config);
  } else {
    connectionOptions.sslmode = 'disable';
  }
  const connection = await pgwire.pgconnect(connectionOptions as pgwire.PgConnectOptions);
  // HACK: Replace row decoding with our own implementation
  (connection as any)._recvDataRow = _recvDataRow;
  return connection;
}

function _recvDataRow(this: any, _message: any, row: Uint8Array[], batch: any) {
  for (let i = 0; i < this._rowColumns.length; i++) {
    const valbuf = row[i];
    if (valbuf == null) {
      continue;
    }
    const { binary, typeOid } = this._rowColumns[i];
    // TODO avoid this._clientTextDecoder.decode for bytea
    row[i] = binary
      ? // do not valbuf.slice() because nodejs Buffer .slice does not copy
        // TODO but we not going to receive Buffer here ?
        Uint8Array.prototype.slice.call(valbuf)
      : PgType.decode(this._rowTextDecoder.decode(valbuf), typeOid);
  }
  batch.rows.push(row);
}

export interface PgPoolOptions {
  /**
   * Maximum number of open connections.
   *
   * Once this limit is reached, queries are queued.
   *
   * Defaults to 5.
   */
  maxSize?: number;

  /**
   * Idle timeout in ms before a connection is closed.
   */
  idleTimeout?: number | undefined;
}

/**
 * Open a postgres pool.
 *
 * The pool cannot be used for transactions, and cannot be used for logical replication.
 */
export function connectPgWirePool(config: PgWireConnectionOptions, options?: PgPoolOptions) {
  const idleTimeout = options?.idleTimeout;
  const maxSize = options?.maxSize ?? 5;

  let connectionOptions: Mutable<ExtendedPgwireOptions> = {
    application_name: 'PowerSync',

    // tlsOptions below contains the original hostname
    hostname: config.resolved_ip ?? config.hostname,
    port: config.port,
    database: config.database,

    user: config.username,
    password: config.password,

    _poolSize: maxSize,
    _poolIdleTimeout: idleTimeout
  };

  if (config.sslmode != 'disable') {
    connectionOptions.sslmode = 'require';

    // HACK: Not standard pgwire options
    // Just the easiest way to pass on our config to pgwire_node.js
    connectionOptions.sslrootcert = tlsOptions(config);
  } else {
    connectionOptions.sslmode = 'disable';
  }

  const pool = pgwire.pgpool(connectionOptions as pgwire.PgConnectOptions);
  const originalGetConnection = (pool as any)._getConnection;
  (pool as any)._getConnection = function (this: any) {
    const con = originalGetConnection.call(this);
    // HACK: Replace row decoding with our own implementation
    (con as any)._recvDataRow = _recvDataRow;
    return con;
  };
  return pool;
}

/**
 * Hack: sslrootcert is passed through as-is to pgwire_node.
 *
 * We use that to pass in custom TLS options, without having to modify pgwire itself.
 */
export interface ExtendedPgwireOptions extends pgwire.PgConnectKnownOptions {
  sslrootcert?: false | tls.ConnectionOptions;
}

export function lsnMakeComparable(text: string) {
  const [h, l] = text.split('/');
  return h.padStart(8, '0') + '/' + l.padStart(8, '0');
}

/**
 * Convert a postgres timestamptz to a SQLite-compatible/normalized timestamp.
 *
 * We output in the format:
 * YYYY-MM-DD HH:MM:SSZ
 * YYYY-MM-DD HH:MM:SS.SSSZ
 *
 * Precision is preserved from the source string, with the exception of ".000".
 *
 * We have specific exceptions for -infinity and infinity.
 */
export function timestamptzToSqlite(source?: string) {
  if (source == null) {
    return null;
  }
  // Make compatible with SQLite
  const match = /^([\d\-]+) ([\d:]+)(\.\d+)?([+-][\d:]+)$/.exec(source);
  if (match == null) {
    if (source == 'infinity') {
      return '9999-12-31 23:59:59Z';
    } else if (source == '-infinity') {
      return '0000-01-01 00:00:00Z';
    } else {
      return null;
    }
  }
  const [_, date, time, precision, timezone] = match as any;
  // datefns.parseISO() supports some dates that new Date() doesn't, e.g. '0022-09-08 21:19:28+00'
  const parsed = datefns.parseISO(`${date}T${time}${timezone}`);
  // Some dates are outside the range of JS and/or SQLite, e.g. 10022-09-08 21:19:28+00.
  // Just fallback to null for now.
  // We may consider preserving the source value in these cases.
  if (isNaN(parsed.getTime())) {
    return null;
  }
  const text = parsed.toISOString().replace('T', ' ').replace('.000', '').replace('Z', '');

  return `${text}${precision ?? ''}Z`;
}

/**
 * For timestamp without timezone, we keep it mostly as-is.
 *
 * We make specific exceptions for 'infinity' and '-infinity'.
 *
 * https://www.postgresql.org/docs/current/datatype-datetime.html#DATATYPE-DATETIME-SPECIAL-VALUES
 */
export function timestampToSqlite(source?: string) {
  if (source == null) {
    return null;
  }
  if (source == 'infinity') {
    return '9999-12-31 23:59:59';
  } else if (source == '-infinity') {
    return '0000-01-01 00:00:00';
  } else {
    return source;
  }
}
/**
 * For date, we keep it mostly as-is.
 *
 * We make specific exceptions for 'infinity' and '-infinity'.
 *
 * https://www.postgresql.org/docs/current/datatype-datetime.html#DATATYPE-DATETIME-SPECIAL-VALUES
 */
export function dateToSqlite(source?: string) {
  if (source == null) {
    return null;
  }
  if (source == 'infinity') {
    return '9999-12-31';
  } else if (source == '-infinity') {
    return '0000-01-01';
  } else {
    return source;
  }
}

/**
 * PgResult rows are arrays of values.
 *
 * This converts it to objects.
 */
export function pgwireRows(rs: pgwire.PgResult): Record<string, any>[] {
  const columns = rs.columns;
  return rs.rows.map((row) => {
    let r: Record<string, any> = {};
    for (let i = 0; i < columns.length; i++) {
      const c = columns[i];
      r[c.name] = row[i];
    }
    return r;
  });
}
