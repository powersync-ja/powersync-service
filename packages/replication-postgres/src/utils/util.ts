import * as pgwire from '@powersync/service-jpgwire';
import { replication } from '@powersync/service-core';

import { ReplicaIdentityResult } from '../types.js';
import { retriedQuery } from './pgwire_utils.js';

export function getReplicaIdColumns(relation: pgwire.PgoutputRelation): replication.ReplicationColumn[] {
  if (relation.replicaIdentity == 'nothing') {
    return [];
  } else {
    return relation.columns.filter((c) => (c.flags & 0b1) != 0).map((c) => ({ name: c.name, typeOid: c.typeOid }));
  }
}

export function getRelId(source: pgwire.PgoutputRelation): number {
  // Source types are wrong here
  const relId = (source as any).relationOid as number;
  if (relId == null || typeof relId != 'number') {
    throw new Error(`No relation id!`);
  }
  return relId;
}

export function getPgOutputRelation(source: pgwire.PgoutputRelation): replication.Relation {
  return {
    name: source.name,
    schema: source.schema,
    relationId: getRelId(source),
    replicaIdentity: source.replicaIdentity,
    replicationColumns: getReplicaIdColumns(source)
  };
}

export async function getPrimaryKeyColumns(
  db: pgwire.PgClient,
  relationId: number,
  mode: 'primary' | 'replident'
): Promise<replication.ReplicationColumn[]> {
  const indexFlag = mode == 'primary' ? `i.indisprimary` : `i.indisreplident`;
  const attrRows = await retriedQuery(db, {
    statement: `SELECT a.attname as name, a.atttypid as typeid, a.attnum as attnum
                                    FROM pg_index i
                                             JOIN pg_attribute a
                                                  ON a.attrelid = i.indrelid
                                                      AND a.attnum = ANY (i.indkey)
                                    WHERE i.indrelid = $1::oid
                                      AND ${indexFlag}
                                      AND a.attnum > 0
                                      ORDER BY a.attnum`,
    params: [{ value: relationId, type: 'int4' }]
  });

  return attrRows.rows.map((row) => {
    return { name: row[0] as string, typeOid: row[1] as number };
  });
}

export async function getAllColumns(db: pgwire.PgClient, relationId: number): Promise<replication.ReplicationColumn[]> {
  const attrRows = await retriedQuery(db, {
    statement: `SELECT a.attname as name, a.atttypid as typeid, a.attnum as attnum
                                    FROM pg_attribute a
                                    WHERE a.attrelid = $1::oid
                                      AND attnum > 0
                                    ORDER BY a.attnum`,
    params: [{ type: 'varchar', value: relationId }]
  });
  return attrRows.rows.map((row) => {
    return { name: row[0] as string, typeOid: row[1] as number };
  });
}

export async function getReplicationIdentityColumns(
  db: pgwire.PgClient,
  relationId: number
): Promise<ReplicaIdentityResult> {
  const rows = await retriedQuery(db, {
    statement: `SELECT CASE relreplident
        WHEN 'd' THEN 'default'
        WHEN 'n' THEN 'nothing'
        WHEN 'f' THEN 'full'
        WHEN 'i' THEN 'index'
     END AS replica_identity
FROM pg_class
WHERE oid = $1::oid LIMIT 1`,
    params: [{ type: 'int8', value: relationId }]
  });
  const idType: string = rows.rows[0]?.[0];
  if (idType == 'nothing' || idType == null) {
    return { replicationIdentity: 'nothing', columns: [] };
  } else if (idType == 'full') {
    return { replicationIdentity: 'full', columns: await getAllColumns(db, relationId) };
  } else if (idType == 'default') {
    return { replicationIdentity: 'default', columns: await getPrimaryKeyColumns(db, relationId, 'primary') };
  } else if (idType == 'index') {
    return { replicationIdentity: 'index', columns: await getPrimaryKeyColumns(db, relationId, 'replident') };
  } else {
    return { replicationIdentity: 'nothing', columns: [] };
  }
}
