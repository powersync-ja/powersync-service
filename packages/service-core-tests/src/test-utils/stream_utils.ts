import { OplogEntry } from '@/util/protocol-types.js';
import { JSONBig } from '@powersync/service-jsonbig';

export function putOp(table: string, data: Record<string, any>): Partial<OplogEntry> {
  return {
    op: 'PUT',
    object_type: table,
    object_id: data.id,
    data: JSONBig.stringify(data)
  };
}

export function removeOp(table: string, id: string): Partial<OplogEntry> {
  return {
    op: 'REMOVE',
    object_type: table,
    object_id: id
  };
}

export function compareIds(a: OplogEntry, b: OplogEntry) {
  return a.object_id!.localeCompare(b.object_id!);
}

export async function oneFromAsync<T>(source: Iterable<T> | AsyncIterable<T>): Promise<T> {
  const items: T[] = [];
  for await (const item of source) {
    items.push(item);
  }
  if (items.length != 1) {
    throw new Error(`One item expected, got: ${items.length}`);
  }
  return items[0];
}

export async function fromAsync<T>(source: Iterable<T> | AsyncIterable<T>): Promise<T[]> {
  const items: T[] = [];
  for await (const item of source) {
    items.push(item);
  }
  return items;
}
