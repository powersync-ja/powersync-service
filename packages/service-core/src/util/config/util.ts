import * as urijs from 'uri-js';

export function getSupabaseJwksUrl(connection: any) {
  if (connection == null) {
    return null;
  } else if (connection.type != 'postgresql') {
    return null;
  }

  let hostname: string | undefined = connection.hostname;
  if (hostname == null && typeof connection.uri == 'string') {
    hostname = urijs.parse(connection.uri).host;
  }
  if (hostname == null) {
    return null;
  }

  const match = /db.(\w+).supabase.co/.exec(hostname);
  if (match == null) {
    return null;
  }
  const projectId = match[1];

  return `https://${projectId}.supabase.co/auth/v1/.well-known/jwks.json`;
}
