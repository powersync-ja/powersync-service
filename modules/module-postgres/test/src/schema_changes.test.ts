import { compareIds, putOp, removeOp } from '@core-tests/stream_utils.js';
import { reduceBucket } from '@powersync/service-core';
import { setTimeout } from 'node:timers/promises';
import { describe, expect, test } from 'vitest';
import { INITIALIZED_MONGO_STORAGE_FACTORY, StorageFactory } from './util.js';
import { WalStreamTestContext } from './wal_stream_utils.js';

describe('schema changes', { timeout: 20_000 }, function () {
  defineTests(INITIALIZED_MONGO_STORAGE_FACTORY);
});

const BASIC_SYNC_RULES = `
bucket_definitions:
  global:
    data:
      - SELECT id, * FROM "test_data"
`;

const PUT_T1 = putOp('test_data', { id: 't1', description: 'test1' });
const PUT_T2 = putOp('test_data', { id: 't2', description: 'test2' });
const PUT_T3 = putOp('test_data', { id: 't3', description: 'test3' });

const REMOVE_T1 = removeOp('test_data', 't1');
const REMOVE_T2 = removeOp('test_data', 't2');

function defineTests(factory: StorageFactory) {
  test('re-create table', async () => {
    await using context = await WalStreamTestContext.open(factory);

    // Drop a table and re-create it.
    await context.updateSyncRules(BASIC_SYNC_RULES);
    const { pool } = context;

    await pool.query(`DROP TABLE IF EXISTS test_data`);
    await pool.query(`CREATE TABLE test_data(id text primary key, description text)`);
    await pool.query(`INSERT INTO test_data(id, description) VALUES('t1', 'test1')`);

    await context.replicateSnapshot();
    context.startStreaming();

    await pool.query(`INSERT INTO test_data(id, description) VALUES('t2', 'test2')`);

    await pool.query(
      { statement: `DROP TABLE test_data` },
      { statement: `CREATE TABLE test_data(id text primary key, description text)` },
      { statement: `INSERT INTO test_data(id, description) VALUES('t3', 'test3')` }
    );

    const data = await context.getBucketData('global[]');

    // Initial inserts
    expect(data.slice(0, 2)).toMatchObject([PUT_T1, PUT_T2]);

    // Truncate - order doesn't matter
    expect(data.slice(2, 4).sort(compareIds)).toMatchObject([REMOVE_T1, REMOVE_T2]);

    expect(data.slice(4)).toMatchObject([
      // Snapshot insert
      PUT_T3,
      // Replicated insert
      // We may eventually be able to de-duplicate this
      PUT_T3
    ]);
  });

  test('add table', async () => {
    await using context = await WalStreamTestContext.open(factory);
    // Add table after initial replication
    await context.updateSyncRules(BASIC_SYNC_RULES);
    const { pool } = context;

    await context.replicateSnapshot();
    context.startStreaming();

    await pool.query(`CREATE TABLE test_data(id text primary key, description text)`);
    await pool.query(`INSERT INTO test_data(id, description) VALUES('t1', 'test1')`);

    const data = await context.getBucketData('global[]');

    expect(data).toMatchObject([
      // Snapshot insert
      PUT_T1,
      // Replicated insert
      // We may eventually be able to de-duplicate this
      PUT_T1
    ]);
  });

  test('rename table (1)', async () => {
    await using context = await WalStreamTestContext.open(factory);
    const { pool } = context;

    await context.updateSyncRules(BASIC_SYNC_RULES);

    // Rename table not in sync rules -> in sync rules
    await pool.query(`CREATE TABLE test_data_old(id text primary key, description text)`);
    await pool.query(`INSERT INTO test_data_old(id, description) VALUES('t1', 'test1')`);

    await context.replicateSnapshot();
    context.startStreaming();

    await pool.query(
      { statement: `ALTER TABLE test_data_old RENAME TO test_data` },
      // We need an operation to detect the change
      { statement: `INSERT INTO test_data(id, description) VALUES('t2', 'test2')` }
    );

    const data = await context.getBucketData('global[]');

    expect(data.slice(0, 2).sort(compareIds)).toMatchObject([
      // Snapshot insert
      PUT_T1,
      PUT_T2
    ]);
    expect(data.slice(2)).toMatchObject([
      // Replicated insert
      // We may eventually be able to de-duplicate this
      PUT_T2
    ]);
  });

  test('rename table (2)', async () => {
    await using context = await WalStreamTestContext.open(factory);
    // Rename table in sync rules -> in sync rules
    const { pool } = context;

    await context.updateSyncRules(`
    bucket_definitions:
      global:
        data:
          - SELECT id, * FROM "test_data%"
    `);

    await pool.query(`CREATE TABLE test_data1(id text primary key, description text)`);
    await pool.query(`INSERT INTO test_data1(id, description) VALUES('t1', 'test1')`);

    await context.replicateSnapshot();
    context.startStreaming();

    await pool.query(
      { statement: `ALTER TABLE test_data1 RENAME TO test_data2` },
      // We need an operation to detect the change
      { statement: `INSERT INTO test_data2(id, description) VALUES('t2', 'test2')` }
    );

    const data = await context.getBucketData('global[]');

    expect(data.slice(0, 2)).toMatchObject([
      // Initial replication
      putOp('test_data1', { id: 't1', description: 'test1' }),
      // Initial truncate
      removeOp('test_data1', 't1')
    ]);

    expect(data.slice(2, 4).sort(compareIds)).toMatchObject([
      // Snapshot insert
      putOp('test_data2', { id: 't1', description: 'test1' }),
      putOp('test_data2', { id: 't2', description: 'test2' })
    ]);
    expect(data.slice(4)).toMatchObject([
      // Replicated insert
      // We may eventually be able to de-duplicate this
      putOp('test_data2', { id: 't2', description: 'test2' })
    ]);
  });

  test('rename table (3)', async () => {
    await using context = await WalStreamTestContext.open(factory);
    // Rename table in sync rules -> not in sync rules

    const { pool } = context;

    await context.updateSyncRules(BASIC_SYNC_RULES);

    await pool.query(`CREATE TABLE test_data(id text primary key, description text)`);
    await pool.query(`INSERT INTO test_data(id, description) VALUES('t1', 'test1')`);

    await context.replicateSnapshot();
    context.startStreaming();

    await pool.query(
      { statement: `ALTER TABLE test_data RENAME TO test_data_na` },
      // We need an operation to detect the change
      { statement: `INSERT INTO test_data_na(id, description) VALUES('t2', 'test2')` }
    );

    const data = await context.getBucketData('global[]');

    expect(data).toMatchObject([
      // Initial replication
      PUT_T1,
      // Truncate
      REMOVE_T1
    ]);
  });

  test('change replica id', async () => {
    await using context = await WalStreamTestContext.open(factory);
    // Change replica id from default to full
    // Causes a re-import of the table.

    const { pool } = context;
    await context.updateSyncRules(BASIC_SYNC_RULES);

    await pool.query(`CREATE TABLE test_data(id text primary key, description text)`);
    await pool.query(`INSERT INTO test_data(id, description) VALUES('t1', 'test1')`);

    await context.replicateSnapshot();
    context.startStreaming();

    await pool.query(
      { statement: `ALTER TABLE test_data REPLICA IDENTITY FULL` },
      // We need an operation to detect the change
      { statement: `INSERT INTO test_data(id, description) VALUES('t2', 'test2')` }
    );

    const data = await context.getBucketData('global[]');

    expect(data.slice(0, 2)).toMatchObject([
      // Initial inserts
      PUT_T1,
      // Truncate
      REMOVE_T1
    ]);

    // Snapshot - order doesn't matter
    expect(data.slice(2, 4).sort(compareIds)).toMatchObject([PUT_T1, PUT_T2]);

    expect(data.slice(4).sort(compareIds)).toMatchObject([
      // Replicated insert
      // We may eventually be able to de-duplicate this
      PUT_T2
    ]);
  });

  test('change full replica id by adding column', async () => {
    await using context = await WalStreamTestContext.open(factory);
    // Change replica id from full by adding column
    // Causes a re-import of the table.
    // Other changes such as renaming column would have the same effect

    const { pool } = context;
    await context.updateSyncRules(BASIC_SYNC_RULES);

    await pool.query(`CREATE TABLE test_data(id text primary key, description text)`);
    await pool.query(`ALTER TABLE test_data REPLICA IDENTITY FULL`);
    await pool.query(`INSERT INTO test_data(id, description) VALUES('t1', 'test1')`);

    await context.replicateSnapshot();
    context.startStreaming();

    await pool.query(
      { statement: `ALTER TABLE test_data ADD COLUMN other TEXT` },
      { statement: `INSERT INTO test_data(id, description) VALUES('t2', 'test2')` }
    );

    const data = await context.getBucketData('global[]');

    expect(data.slice(0, 2)).toMatchObject([
      // Initial inserts
      PUT_T1,
      // Truncate
      REMOVE_T1
    ]);

    // Snapshot - order doesn't matter
    expect(data.slice(2, 4).sort(compareIds)).toMatchObject([
      putOp('test_data', { id: 't1', description: 'test1', other: null }),
      putOp('test_data', { id: 't2', description: 'test2', other: null })
    ]);

    expect(data.slice(4).sort(compareIds)).toMatchObject([
      // Replicated insert
      // We may eventually be able to de-duplicate this
      putOp('test_data', { id: 't2', description: 'test2', other: null })
    ]);
  });

  test('change default replica id by changing column type', async () => {
    await using context = await WalStreamTestContext.open(factory);
    // Change default replica id by changing column type
    // Causes a re-import of the table.
    const { pool } = context;
    await context.updateSyncRules(BASIC_SYNC_RULES);

    await pool.query(`CREATE TABLE test_data(id text primary key, description text)`);
    await pool.query(`INSERT INTO test_data(id, description) VALUES('t1', 'test1')`);

    await context.replicateSnapshot();
    context.startStreaming();

    await pool.query(
      { statement: `ALTER TABLE test_data ALTER COLUMN id TYPE varchar` },
      { statement: `INSERT INTO test_data(id, description) VALUES('t2', 'test2')` }
    );

    const data = await context.getBucketData('global[]');

    expect(data.slice(0, 2)).toMatchObject([
      // Initial inserts
      PUT_T1,
      // Truncate
      REMOVE_T1
    ]);

    // Snapshot - order doesn't matter
    expect(data.slice(2, 4).sort(compareIds)).toMatchObject([PUT_T1, PUT_T2]);

    expect(data.slice(4).sort(compareIds)).toMatchObject([
      // Replicated insert
      // We may eventually be able to de-duplicate this
      PUT_T2
    ]);
  });

  test('change index id by changing column type', async () => {
    await using context = await WalStreamTestContext.open(factory);
    // Change index replica id by changing column type
    // Causes a re-import of the table.
    // Secondary functionality tested here is that replica id column order stays
    // the same between initial and incremental replication.
    const { pool } = context;
    await context.updateSyncRules(BASIC_SYNC_RULES);

    await pool.query(`CREATE TABLE test_data(id text primary key, description text not null)`);
    await pool.query(`CREATE UNIQUE INDEX i1 ON test_data(description, id)`);
    await pool.query(`ALTER TABLE test_data REPLICA IDENTITY USING INDEX i1`);

    await pool.query(`INSERT INTO test_data(id, description) VALUES('t1', 'test1')`);

    await context.replicateSnapshot();
    context.startStreaming();

    await pool.query(`INSERT INTO test_data(id, description) VALUES('t2', 'test2')`);

    await pool.query(
      { statement: `ALTER TABLE test_data ALTER COLUMN description TYPE varchar` },
      { statement: `INSERT INTO test_data(id, description) VALUES('t3', 'test3')` }
    );

    const data = await context.getBucketData('global[]');

    expect(data.slice(0, 2)).toMatchObject([
      // Initial snapshot
      PUT_T1,
      // Streamed
      PUT_T2
    ]);

    expect(data.slice(2, 4).sort(compareIds)).toMatchObject([
      // Truncate - any order
      REMOVE_T1,
      REMOVE_T2
    ]);

    // Snapshot - order doesn't matter
    expect(data.slice(4, 7).sort(compareIds)).toMatchObject([PUT_T1, PUT_T2, PUT_T3]);

    expect(data.slice(7).sort(compareIds)).toMatchObject([
      // Replicated insert
      // We may eventually be able to de-duplicate this
      PUT_T3
    ]);
  });

  test('add to publication', async () => {
    await using context = await WalStreamTestContext.open(factory);
    // Add table to publication after initial replication
    const { pool } = context;

    await pool.query(`DROP PUBLICATION powersync`);
    await pool.query(`CREATE TABLE test_foo(id text primary key, description text)`);
    await pool.query(`CREATE PUBLICATION powersync FOR table test_foo`);

    const storage = await context.updateSyncRules(BASIC_SYNC_RULES);

    await pool.query(`CREATE TABLE test_data(id text primary key, description text)`);
    await pool.query(`INSERT INTO test_data(id, description) VALUES('t1', 'test1')`);

    await context.replicateSnapshot();
    context.startStreaming();

    await pool.query(`INSERT INTO test_data(id, description) VALUES('t2', 'test2')`);

    await pool.query(`ALTER PUBLICATION powersync ADD TABLE test_data`);
    await pool.query(`INSERT INTO test_data(id, description) VALUES('t3', 'test3')`);

    const data = await context.getBucketData('global[]');

    expect(data.slice(0, 3).sort(compareIds)).toMatchObject([
      // Snapshot insert - any order
      PUT_T1,
      PUT_T2,
      PUT_T3
    ]);

    expect(data.slice(3)).toMatchObject([
      // Replicated insert
      // We may eventually be able to de-duplicate this
      PUT_T3
    ]);

    const metrics = await storage.factory.getStorageMetrics();
    expect(metrics.replication_size_bytes).toBeGreaterThan(0);
  });

  test('add to publication (not in sync rules)', async () => {
    await using context = await WalStreamTestContext.open(factory);
    // Add table to publication after initial replication
    // Since the table is not in sync rules, it should not be replicated.
    const { pool } = context;

    await pool.query(`DROP PUBLICATION powersync`);
    await pool.query(`CREATE TABLE test_foo(id text primary key, description text)`);
    await pool.query(`CREATE PUBLICATION powersync FOR table test_foo`);

    const storage = await context.updateSyncRules(BASIC_SYNC_RULES);

    await pool.query(`CREATE TABLE test_other(id text primary key, description text)`);
    await pool.query(`INSERT INTO test_other(id, description) VALUES('t1', 'test1')`);

    await context.replicateSnapshot();
    context.startStreaming();

    await pool.query(`INSERT INTO test_other(id, description) VALUES('t2', 'test2')`);

    await pool.query(`ALTER PUBLICATION powersync ADD TABLE test_other`);
    await pool.query(`INSERT INTO test_other(id, description) VALUES('t3', 'test3')`);

    const data = await context.getBucketData('global[]');
    expect(data).toMatchObject([]);

    const metrics = await storage.factory.getStorageMetrics();
    expect(metrics.replication_size_bytes).toEqual(0);
  });

  test('replica identity nothing', async () => {
    await using context = await WalStreamTestContext.open(factory);
    // Technically not a schema change, but fits here.
    // Replica ID works a little differently here - the table doesn't have
    // one defined, but we generate a unique one for each replicated row.

    const { pool } = context;
    await context.updateSyncRules(BASIC_SYNC_RULES);

    await pool.query(`CREATE TABLE test_data(id text primary key, description text)`);
    await pool.query(`ALTER TABLE test_data REPLICA IDENTITY NOTHING`);
    await pool.query(`INSERT INTO test_data(id, description) VALUES('t1', 'test1')`);

    await context.replicateSnapshot();
    context.startStreaming();

    await pool.query(`INSERT INTO test_data(id, description) VALUES('t2', 'test2')`);

    // Just as an FYI - cannot update or delete here
    expect(pool.query(`UPDATE test_data SET description = 'test2b' WHERE id = 't2'`)).rejects.toThrow(
      'does not have a replica identity and publishes updates'
    );

    // Testing TRUNCATE is important here - this depends on current_data having unique
    // ids.
    await pool.query(`TRUNCATE TABLE test_data`);

    const data = await context.getBucketData('global[]');

    expect(data.slice(0, 2)).toMatchObject([
      // Initial inserts
      PUT_T1,
      PUT_T2
    ]);

    expect(data.slice(2).sort(compareIds)).toMatchObject([
      // Truncate
      REMOVE_T1,
      REMOVE_T2
    ]);
  });

  test('replica identity default without PK', async () => {
    await using context = await WalStreamTestContext.open(factory);
    // Same as no replica identity
    const { pool } = context;
    await context.updateSyncRules(BASIC_SYNC_RULES);

    await pool.query(`CREATE TABLE test_data(id text, description text)`);
    await pool.query(`INSERT INTO test_data(id, description) VALUES('t1', 'test1')`);

    await context.replicateSnapshot();
    context.startStreaming();

    await pool.query(`INSERT INTO test_data(id, description) VALUES('t2', 'test2')`);

    // Just as an FYI - cannot update or delete here
    expect(pool.query(`UPDATE test_data SET description = 'test2b' WHERE id = 't2'`)).rejects.toThrow(
      'does not have a replica identity and publishes updates'
    );

    // Testing TRUNCATE is important here - this depends on current_data having unique
    // ids.
    await pool.query(`TRUNCATE TABLE test_data`);

    const data = await context.getBucketData('global[]');

    expect(data.slice(0, 2)).toMatchObject([
      // Initial inserts
      PUT_T1,
      PUT_T2
    ]);

    expect(data.slice(2).sort(compareIds)).toMatchObject([
      // Truncate
      REMOVE_T1,
      REMOVE_T2
    ]);
  });

  // Test consistency of table snapshots.
  // Renames a table to trigger a snapshot.
  // To trigger the failure, modify the snapshot implementation to
  // introduce an arbitrary delay (in WalStream.ts):
  //
  //   const rs = await db.query(`select pg_current_wal_lsn() as lsn`);
  //   lsn = rs.rows[0][0];
  //   await new Promise((resolve) => setTimeout(resolve, 100));
  //   await this.snapshotTable(batch, db, result.table);
  test('table snapshot consistency', async () => {
    await using context = await WalStreamTestContext.open(factory);
    const { pool } = context;

    await context.updateSyncRules(BASIC_SYNC_RULES);

    // Rename table not in sync rules -> in sync rules
    await pool.query(`CREATE TABLE test_data_old(id text primary key, num integer)`);
    await pool.query(`INSERT INTO test_data_old(id, num) VALUES('t1', 0)`);
    await pool.query(`INSERT INTO test_data_old(id, num) VALUES('t2', 0)`);

    await context.replicateSnapshot();
    context.startStreaming();

    await pool.query(
      { statement: `ALTER TABLE test_data_old RENAME TO test_data` },
      // This first update will trigger a snapshot
      { statement: `UPDATE test_data SET num = 0 WHERE id = 't2'` }
    );

    // Need some delay for the snapshot to be triggered
    await setTimeout(5);

    let stop = false;

    let failures: any[] = [];

    // This is a tight loop that checks that t2.num >= t1.num
    const p = (async () => {
      let lopid = '';
      while (!stop) {
        const data = await context.getCurrentBucketData('global[]');
        const last = data[data.length - 1];
        if (last == null) {
          continue;
        }
        if (last.op_id != lopid) {
          const reduced = reduceBucket(data);
          reduced.shift();
          lopid = last.op_id;

          const t1 = reduced.find((op) => op.object_id == 't1');
          const t2 = reduced.find((op) => op.object_id == 't2');
          if (t1 && t2) {
            const d1 = JSON.parse(t1.data as string);
            const d2 = JSON.parse(t2.data as string);
            if (d1.num > d2.num) {
              failures.push({ d1, d2 });
            }
          }
        }
      }
    })();

    // We always have t2.num >= t1.num
    for (let i = 1; i <= 20; i++) {
      await pool.query({ statement: `UPDATE test_data SET num = ${i} WHERE id = 't2'` });
    }
    await pool.query({ statement: `UPDATE test_data SET num = 20 WHERE id = 't1'` });

    await context.getBucketData('global[]');
    stop = true;
    await p;

    expect(failures).toEqual([]);
  });
}
