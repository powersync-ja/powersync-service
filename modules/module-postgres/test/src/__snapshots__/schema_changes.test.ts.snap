// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`schema changes > mongodb storage > add to publication (not in sync rules) 1`] = `0`;

exports[`schema changes > postgres storage > add to publication (not in sync rules) 1`] = `16384`;
